### PR TITLE
Add missing fields for tournament

### PR DESCRIPTION
**Pull request by istar410**
Add some missing fields like the number of player per team or the number of validated teams.

### CODE REVIEW

**Review by KwikKill — APPROVED** [review]
nice :ok_hand:

**Comment by Lymkwi** [comment]
Why are there migrations?

**Comment by KwikKill** [comment]
> Why are there migrations?

Migrations are not generated by the prod environment atm (I don't even know why) and conflict migrations are hard to handle when using a filled database.
The easiest way to solve theses problems is to commit the migrations (https://paulonteri.com/thoughts/should-you-commit-migrations).

**Comment by Lugrim** [comment]
> The easiest way to solve theses problems is to commit the migrations (https://paulonteri.com/thoughts/should-you-commit-migrations).

Indeed, but migrations should only be committed between 2 production deployment, not in PRs. The review *will* ask for model changes, so we will have dozens of useless migrations adding, patching, reverting changes, etc.

**Comment by KwikKill** [comment]
> > The easiest way to solve theses problems is to commit the migrations (https://paulonteri.com/thoughts/should-you-commit-migrations).
> 
> Indeed, but migrations should only be committed between 2 production deployment, not in PRs. The review _will_ ask for model changes, so we will have dozens of useless migrations adding, patching, reverting changes, etc.

Ok, I agree with that, that was a misconception of mine. I will commit a fix in this PR soon.

File: insalan/tournament/models.py [hunk 1]
```diff
@@ -615,4 +640,27 @@ def clean(self):
...
             )
 
+class Caster(models.Model):
+    
+    pseudo = models.CharField(
+        max_length=42,
+        null=False,
+        blank=False,
+        verbose_name=_("Pseudo du casteur")
+    )
```
Comment (Lugrim) [inline insalan/tournament/models.py:650]: I feel like "pseudo" still sound very French

**Review by Lugrim — CHANGES_REQUESTED** [review]
Migrations should be removed.

Also, I just figured out we never update the `validated` state of a team upon updating it, maybe we should ensure that we can't remove players who already paid and handle refunds (set to False the validated state if we go below the threshold)?

File: insalan/tournament/models.py (hunk 1 above)
Comment (istar410) [reply insalan/tournament/models.py:650]: What would you suggest to replace ?

Comment (KwikKill) [reply insalan/tournament/models.py:650]: "username" or "in_game_name" maybe ?

Comment (Lugrim) [reply insalan/tournament/models.py:650]: Not for casters, simply something like "name"?

Comment (istar410) [reply insalan/tournament/models.py:650]: "name" sounds good to me.

File: insalan/tournament/models.py [hunk 2]
```diff
@@ -615,4 +640,27 @@ def clean(self):
...
+    image = models.FileField(
+        verbose_name=_("Photo de profil"),
+        blank=True,
+        null=True,
+        upload_to="profile-pictures",
+        validators=[
+            FileExtensionValidator(allowed_extensions=["png", "jpg", "jpeg", "svg"])
+        ],
+    )
```
[the remark this answers is not in the document]
Comment (istar410) [reply insalan/tournament/models.py:659]: I have added an url field.

**Review by KwikKill — APPROVED** [review]

**Review by KwikKill — APPROVED** [review]

**Review by Lymkwi — CHANGES_REQUESTED** [review]
I'm sorry to say, but testing this code makes me extremely worried about the state of our backend logic, notably because I hit bugs introduced by *other PRs* while trying to run it.

Here are the things I noticed that could be problems:
- The code path to deny team creation if there are already enough validated doesn't work on the admin panel
- It's also not working on the POST path to `/tournament/team/`
- What of multiple teams awaiting to be validated? What happens if the last one possible becomes valid? There is no modification of the validation refresh code to account for the fact that there could already be enough validated teams
- You can create teams to an unannounced tournament! I did not know that. We should probably fix that
- If you manage to trigger the code path to fail on multiple teams it crashes django (see comments)
- You should probably add `players_per_team` to the display fields of Game's admin view
- The player registration form in the browsable API does not contain any password field, probably because the serializer for Player does not mention it (because it's not part of a Player). I assume the same thing happens for manager.
- I'm gonna assume nobody tested all of this, because I had to fix other code not from this PR for the code to be triggered
- If you try and register one more person than possible in a team, you will hit the "This person is already registered for this event" validator, rather than something like "sorry, too many registrations"

File: insalan/tournament/views.py [hunk 3]
```diff
@@ -278,6 +278,16 @@ def post(self, request, *args, **kwargs):
                 }
             )
 
+        trnm = Tournament.objects.get(pk=data["tournament"])
+        if trnm.get_validated_teams() == trnm.get_maxTeam():
+            return PermissionDenied(
```
[the remark this answers is not in the document]
Comment (istar410) [reply insalan/tournament/views.py:283]: Yes, I will raise it.

**Comment by Lymkwi** [comment]
Logic-wise it all works as I'd expect it to, without tests but, you know, we can add those later.

Now it's time to see if I can nitpick some code.

**Review by Lymkwi — APPROVED** [review]
lgtm, I think we finally covered everything.

File: insalan/tournament/admin.py [hunk 4]
```diff
@@ -53,8 +53,8 @@ class TeamAdmin(admin.ModelAdmin):
 class PlayerAdmin(admin.ModelAdmin):
     """Admin handler for Player Registrations"""
 
-    list_display = ("id", "user", "team", "payment_status", "ticket")
-    search_fields = ["user", "team", "payment_status"]
+    list_display = ("id", "user", "pseudo", "team", "payment_status", "ticket")
```
[the remark this answers is not in the document]
Comment (KwikKill) [reply insalan/tournament/admin.py:284]: You want my death but you're right, i'll do it now.
image

File: insalan/tournament/payment.py [hunk 5]
```diff
@@ -23,7 +23,7 @@ def fetch_registration(product, user):
         """
         Fetch a registration for a user and product.
 
-        Returns a tuple (reg, is_manager), which is pretty explicit. Raises
+        Returns a tuple (reg, is_manager, is_substitute), which is pretty explicit. Raises
```
Comment (Lugrim) [inline insalan/tournament/payment.py:26]: Wouldn't it make more sense to use an "enum" than two booleans?

**Review by Lugrim — APPROVED** [review]
Looks goodenough to merge to me. Some nitpicking on code, but nothing needed right now.

Also, we may want to refactor (later) to have superclasses of anything Player/Manager/Substitute related, it looks quite a bit of code is reused

File: insalan/tournament/payment.py (hunk 5 above)
Comment (KwikKill) [reply insalan/tournament/payment.py:26]: There's a lot of duplicated codes in this file. I think we keep it like that for now and we'll refactor everything later.

File: insalan/tournament/admin.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Lymkwi) [reply insalan/tournament/admin.py]: Goodness me. Good luck.

**Comment by Lymkwi** [comment]
Wait, actually, there's one thing wrong: you accidentally added a migration

**Comment by KwikKill** [comment]
> Wait, actually, there's one thing wrong: you accidentally added a migration

Oh, yes, mb

**Comment by Lymkwi** [comment]
> > Wait, actually, there's one thing wrong: you accidentally added a migration
> 
> Oh, yes, mb

Wait, actually it turns out they might come from the production environment?

**Comment by KwikKill** [comment]
> > > Wait, actually, there's one thing wrong: you accidentally added a migration
> > 
> > 
> > Oh, yes, mb
> 
> Wait, actually it turns out they might come from the production environment? A ton of them were pushed earlier this month. Maybe we can handle that elsewhere anyways.

The migration file is the one currently in migration, yes, but the pseudo field rename should not be there, it should be commit later for the prod deployment preparation.

**Pull request merged**